### PR TITLE
Session resume broken switching contexts

### PR DESCRIPTION
**Pull request by tmshort**
When an SSL's context is swtiched from a ticket-enabled context to
a ticket-disabled context in the servername callback, no session-id
is generated, so the session can't be resumed.

If a servername callback changes the SSL_OP_NO_TICKET option, check
to see if it's changed to disable, and whether a session ticket is
expected (i.e. the client indicated ticket support and the SSL had
tickets enabled at the time), and whether we already have a previous
session (i.e. s->hit is set).

In this case, clear the ticket-expected flag and create a new session
that has a session-id associated with it.

If the SSL hit (resumed) and switched to a ticket-disabled context,
assume that the resumption was via session-id, and don't bother to
create a new session.

Before this fix, the updated unit-tests in 06-sni-ticket.conf would
fail test #4 (server1 = SNI, server2 = no SNI).

### CODE REVIEW

**Comment by ekasper** [comment]
(drive-by) Haven't though about the gist of it yet, but this needs a test.

**Comment by tmshort** [comment]
Unit tests added.

**Comment by tmshort** [comment]
ping?

**Comment by FdaSilvaYY** [comment]
According to GitHub, this PR needs a rebase.

> Conflicting files
> ssl/t1_lib.c
> test/handshake_helper.h
> test/ssl-tests/06-sni-ticket.conf.in
> test/ssl_test_ctx.h

**Comment by tmshort** [comment]
Rebased.

File: test/handshake_helper.c [hunk 1]
```diff
@@ -1118,6 +1118,8 @@ static HANDSHAKE_RESULT *do_handshake_internal(
     handshake_status_t status = HANDSHAKE_RETRY;
     const unsigned char* tick = NULL;
     size_t tick_len = 0;
+    const unsigned char* sess_id= NULL;
```
Comment (FdaSilvaYY) [inline test/handshake_helper.c]: nit: `sess = NULL;`

File: ssl/ssl_sess.c [hunk 2]
```diff
@@ -300,13 +300,91 @@ static int def_generate_session_id(const SSL *ssl, unsigned char *id,
...
+    if (tmp == 0 || tmp > ss->session_id_length) {
+        /* The callback set an illegal length */
+        SSLerr(SSL_F_SSL_GENERATE_SESSION_ID,
+               SSL_R_SSL_SESSION_ID_HAS_BAD_LENGTH);
+        return (0);
+    }
+    ss->session_id_length = tmp;
+    /* Finally, check for a conflict */
+    if (SSL_has_matching_session_id(s, ss->session_id,
+                                    (unsigned int)ss->session_id_length)) {
```
Comment (FdaSilvaYY) [inline ssl/ssl_sess.c]: Nit: space after cast ?
`(unsigned int) ss->session_id_length)) {`

File: test/handshake_helper.c (hunk 1 above)
Comment (tmshort) [inline test/handshake_helper.c]: fixed.

File: ssl/ssl_sess.c (hunk 2 above)
Comment (tmshort) [inline ssl/ssl_sess.c]: Most casts I see for (unsigned int) and (int) do not have a space after the cast, so leaving as-is.

**Comment by tmshort** [comment]
ping?

**Comment by tmshort** [comment]
Rebased.

**Comment by tmshort** [comment]
ping @openssl/cmo

File: ssl/ssl_sess.c [hunk 3]
```diff
@@ -308,13 +308,91 @@ static int def_generate_session_id(const SSL *ssl, unsigned char *id,
...
+{
+    unsigned int tmp;
+    GEN_SESSION_CB cb = def_generate_session_id;
+
+    switch (s->version) {
+    case SSL3_VERSION:
+    case TLS1_VERSION:
+    case TLS1_1_VERSION:
+    case TLS1_2_VERSION:
+    case TLS1_3_VERSION:
```
Comment (mattcaswell) [inline ssl/ssl_sess.c]: TLSv1.3 doesn't have session ids, so is this appropriate?

File: ssl/ssl_sess.c [hunk 4]
```diff
@@ -308,13 +308,91 @@ static int def_generate_session_id(const SSL *ssl, unsigned char *id,
...
+     *     however, because of the lookahead, it already knows
+     *     whether a ticket is expected or not.
+     *
+     * (b) statem_clnt.c calls ssl_get_new_session() before parsing
+     *     ServerHello extensions, and before recording the session
+     *     ID received from the server, so this block is a noop.
+     */
+    if (s->ext.ticket_expected) {
+        ss->session_id_length = 0;
+        return (1);
```
Comment (mattcaswell) [inline ssl/ssl_sess.c]: no ()

File: ssl/ssl_sess.c [hunk 5]
```diff
@@ -308,13 +308,91 @@ static int def_generate_session_id(const SSL *ssl, unsigned char *id,
...
+    case TLS1_2_VERSION:
+    case TLS1_3_VERSION:
+    case DTLS1_BAD_VER:
+    case DTLS1_VERSION:
+    case DTLS1_2_VERSION:
+        ss->session_id_length = SSL3_SSL_SESSION_ID_LENGTH;
+        break;
+    default:
+        SSLerr(SSL_F_SSL_GENERATE_SESSION_ID, SSL_R_UNSUPPORTED_SSL_VERSION);
+        return (0);
```
Comment (mattcaswell) [inline ssl/ssl_sess.c]: no ()

File: ssl/ssl_sess.c [hunk 6]
```diff
@@ -308,13 +308,91 @@ static int def_generate_session_id(const SSL *ssl, unsigned char *id,
...
+    }
+    /*
+     * Don't allow the callback to set the session length to zero. nor
+     * set it higher than it was.
+     */
+    if (tmp == 0 || tmp > ss->session_id_length) {
+        /* The callback set an illegal length */
+        SSLerr(SSL_F_SSL_GENERATE_SESSION_ID,
+               SSL_R_SSL_SESSION_ID_HAS_BAD_LENGTH);
+        return (0);
```
Comment (mattcaswell) [inline ssl/ssl_sess.c]: () etc

File: ssl/ssl_sess.c [hunk 7]
```diff
@@ -329,96 +407,11 @@ int ssl_get_new_session(SSL *s, int session)
...
-            return (0);
-        }
-        ss->session_id_length = tmp;
-        /* Finally, check for a conflict */
-        if (SSL_has_matching_session_id(s, ss->session_id,
-                                        (unsigned int)ss->session_id_length)) {
-            SSLerr(SSL_F_SSL_GET_NEW_SESSION, SSL_R_SSL_SESSION_ID_CONFLICT);
+        if (!ssl_generate_session_id(s, ss)) {
             SSL_SESSION_free(ss);
             return (0);
```
Comment (mattcaswell) [inline ssl/ssl_sess.c]: ()

File: ssl/statem/extensions.c [hunk 8]
```diff
@@ -830,6 +831,34 @@ static int final_server_name(SSL *s, unsigned int context, int sent,
         ret = s->session_ctx->ext.servername_cb(s, &altmp,
                                        s->session_ctx->ext.servername_arg);
 
+    /*
+     * If we're expecting to send a ticket, and tickets were previously enabled,
+     * and now tickets are disabled, then turn off expected ticket.
+     * Also, if this is not a resumption, create a new session ID
+     */
+    if (s->ext.ticket_expected && was_ticket &&
```
Comment (mattcaswell) [inline ssl/statem/extensions.c]: Does this block only apply if ret == SSL_TLSEXT_ERR_OK?

File: ssl/statem/extensions.c [hunk 9]
```diff
@@ -830,6 +831,34 @@ static int final_server_name(SSL *s, unsigned int context, int sent,
...
+        (SSL_get_options(s) & SSL_OP_NO_TICKET) != 0) {
+        s->ext.ticket_expected = 0;
+        if (!s->hit) {
+            SSL_SESSION* ss = SSL_get_session(s);
```
Comment (mattcaswell) [inline ssl/statem/extensions.c]: Blank line after declaration

File: ssl/ssl_sess.c (hunk 3 above)
Comment (tmshort) [inline ssl/ssl_sess.c]: The original code (deleted) included TLS1_3_VERSION.

File: ssl/ssl_sess.c (hunk 5 above)
Comment (tmshort) [inline ssl/ssl_sess.c]: Parens were from the original code, and retianed.

File: ssl/ssl_sess.c (hunk 3 above)
Comment (kaduk) [inline ssl/ssl_sess.c]: IIRC, we still generate a session ID in this field to have a unique internal handle and simplify other code paths.  Someone (maybe Ilari or davidben?) pointed out security issues with the previous state of affairs w.r.t TLS 1.3

File: ssl/statem/extensions.c (hunk 8 above)
Comment (tmshort) [inline ssl/statem/extensions.c]: Yes.

File: apps/progs.h [hunk 10]
```diff
@@ -2,7 +2,7 @@
  * WARNING: do not edit!
  * Generated by apps/progs.pl
  *
- * Copyright 1995-2016 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 1995-2017 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (tmshort) [inline apps/progs.h]: `make update` updates this now.

Comment (richsalz) [inline apps/progs.h]: just recently fixed, but waiting for a push to go through.

**Comment by tmshort** [comment]
@mattcaswell are you ok with this now?

**Comment by tmshort** [comment]
Rebased.

**Comment by tmshort** [comment]
ping @mattcaswell ?

**Comment by tmshort** [comment]
ping @mattcaswell @openssl/omc ?

**Comment by tmshort** [comment]
Ping @openssl/committers ?

**Comment by tmshort** [comment]
Thank you!

**Comment by paulidale** [comment]
Merged and closed.

**Pull request closed without merging**